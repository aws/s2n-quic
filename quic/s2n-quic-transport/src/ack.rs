// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

pub mod interest {
    #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
    pub enum Interest {
        None,
        #[allow(dead_code)]
        Immediate,
    }

    impl Interest {
        #[inline]
        pub fn is_none(self) -> bool {
            matches!(self, Interest::None)
        }
    }

    pub trait Provider {
        fn ack_interest<Q: Query>(&self, query: &mut Q) -> Result;

        #[inline]
        fn has_ack_interest(&self) -> bool {
            let mut query = HasAckInterestQuery;
            self.ack_interest(&mut query).is_err()
        }
    }

    pub trait Query {
        fn on_interest(&mut self, interest: Interest) -> Result;
    }

    pub struct HasAckInterestQuery;

    impl Query for HasAckInterestQuery {
        #[inline]
        fn on_interest(&mut self, interest: Interest) -> Result {
            if interest.is_none() {
                Ok(())
            } else {
                // If we've got anything other than `None` then bail since we now have an answer
                Err(QueryBreak)
            }
        }
    }

    pub struct QueryBreak;

    pub type Result<T = (), E = QueryBreak> = core::result::Result<T, E>;
}
