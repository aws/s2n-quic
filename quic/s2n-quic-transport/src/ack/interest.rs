// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum Interest {
    None,
    #[allow(dead_code)]
    Immediate,
}

impl Interest {
    #[inline]
    pub fn is_none(self) -> bool {
        matches!(self, Interest::None)
    }
}

pub trait Provider {
    fn ack_interest<Q: Query>(&self, query: &mut Q) -> Result;

    #[inline]
    fn has_ack_interest(&self) -> bool {
        let mut query = HasAckInterestQuery;
        self.ack_interest(&mut query).is_err()
    }
}

pub trait Query {
    fn on_interest(&mut self, interest: Interest) -> Result;
}

pub struct HasAckInterestQuery;

impl Query for HasAckInterestQuery {
    #[inline]
    fn on_interest(&mut self, interest: Interest) -> Result {
        if interest.is_none() {
            Ok(())
        } else {
            // If we've got anything other than `None` then bail since we now have an answer
            Err(QueryBreak)
        }
    }
}

pub struct QueryBreak;

pub type Result<T = (), E = QueryBreak> = core::result::Result<T, E>;

#[cfg(test)]
mod test {
    use super::{Provider, Query, *};

    #[test]
    fn has_ack_interest() {
        assert!(!Foo.has_ack_interest());
        assert!(Bar.has_ack_interest());
        assert!(Buzz { foo: Foo, bar: Bar }.has_ack_interest());
    }

    struct Foo;
    impl Provider for Foo {
        fn ack_interest<Q: Query>(&self, query: &mut Q) -> super::Result {
            query.on_interest(Interest::None)
        }
    }
    struct Bar;
    impl Provider for Bar {
        fn ack_interest<Q: Query>(&self, query: &mut Q) -> super::Result {
            query.on_interest(Interest::Immediate)
        }
    }
    struct Buzz {
        foo: Foo,
        bar: Bar,
    }
    impl Provider for Buzz {
        fn ack_interest<Q: Query>(&self, query: &mut Q) -> super::Result {
            self.foo.ack_interest(query)?;
            self.bar.ack_interest(query)?;
            query.on_interest(Interest::None)
        }
    }
}
