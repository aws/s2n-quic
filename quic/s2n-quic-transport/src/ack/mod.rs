// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

pub use ack_manager::*;
pub use s2n_quic_core::ack::*;

mod ack_eliciting_transmission;
mod ack_manager;
mod ack_transmission_state;

#[cfg(test)]
mod tests;
