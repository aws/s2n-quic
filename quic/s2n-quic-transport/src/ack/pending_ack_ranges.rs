// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{ack::ack_ranges::AckRanges, path};
use core::time::Duration;
use s2n_quic_core::{
    frame::ack::EcnCounts,
    packet::number::{PacketNumber, PacketNumberRange},
};

/// Stores aggregated ACK info for delayed processing
#[derive(Clone, Debug, Default)]
pub struct PendingAckRanges {
    ack_ranges: AckRanges,
    ecn_counts: EcnCounts,
    ack_delay: Duration,
    /// The path for which to aggregate ACKs
    pub current_active_path: Option<path::Id>,
}

impl PendingAckRanges {
    /// Extend with a packet number range; dropping smaller values if needed
    #[inline]
    pub fn extend(
        &mut self,
        acked_packets: impl Iterator<Item = PacketNumberRange>,
        ecn_counts: Option<EcnCounts>,
        ack_delay: Duration,
    ) -> Result<(), ()> {
        debug_assert!(
            self.current_active_path.is_some(),
            "active path should be set prior to inserting acks"
        );

        if let Some(ecn_counts) = ecn_counts {
            self.ecn_counts = self.ecn_counts.max(ecn_counts);
        }
        // TODO: at the moment only a single payload(single delivery) worth of ACKs is
        // batched for processing. This means that its acceptable to take the max
        // ack_delay value.
        //
        // Once multiple payloads are stored/batched, multiple ack_delays might need to
        // be stored.
        self.ack_delay = self.ack_delay.max(ack_delay);

        let mut did_insert = true;
        for range in acked_packets {
            did_insert &= self.ack_ranges.insert_packet_number_range(range).is_ok()
        }

        match did_insert {
            true => Ok(()),
            false => Err(()),
        }
    }

    /// Returns an iterator over all values in the `AckRanges`
    #[inline]
    pub fn iter(&self) -> impl Iterator<Item = PacketNumberRange> + '_ {
        debug_assert!(
            self.current_active_path.is_some(),
            "active path should be set prior to processing acks"
        );

        self.ack_ranges
            .inclusive_ranges()
            .into_iter()
            .map(|ack_range| PacketNumberRange::new(*ack_range.start(), *ack_range.end()))
    }

    /// Returns `EcnCounts` aggregated over all the pending ACKs
    #[inline]
    pub fn ecn_counts(&self) -> Option<EcnCounts> {
        debug_assert!(
            self.current_active_path.is_some(),
            "active path should be set prior to processing acks"
        );

        if self.ack_ranges.is_empty() {
            None
        } else {
            Some(self.ecn_counts)
        }
    }

    /// Returns the ACK delay associated with all the pending ACKs
    #[inline]
    pub fn ack_delay(&self) -> Duration {
        debug_assert!(
            self.current_active_path.is_some(),
            "active path should be set prior to processing acks"
        );

        self.ack_delay
    }

    /// Set the current active path for which to aggregate ACKs
    #[inline]
    pub fn set_active_path(&mut self, path_id: path::Id) {
        self.current_active_path = Some(path_id)
    }

    /// Returns the largest `PacketNumber` stored in the AckRanges.
    ///
    /// If no items are present in the set, `None` is returned.
    pub fn max_value(&self) -> Option<PacketNumber> {
        debug_assert!(
            self.current_active_path.is_some(),
            "active path should be set prior to processing acks"
        );

        self.ack_ranges.max_value()
    }

    /// Clear the aggregated ACK information
    #[inline]
    pub fn reset_aggregate_info(&mut self) {
        self.ack_ranges.clear();
        self.ecn_counts = EcnCounts::default();
        self.ack_delay = Duration::default();
    }

    /// Re-initialize all fields.
    ///
    /// Should be called at the end of a processing round to clear all
    /// data. Resets aggregated ack information and the current_active_path.
    #[inline]
    pub fn reset(&mut self) {
        debug_assert!(
            self.current_active_path.is_some(),
            "reset called more than once"
        );
        self.reset_aggregate_info();
        self.current_active_path = None;
    }

    /// Returns if ack ranges are being tracked
    #[inline]
    pub fn is_empty(&self) -> bool {
        self.ack_ranges.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::{super::tests::packet_numbers_iter, *};
    use bolero::check;
    use s2n_quic_core::{
        frame::ack::EcnCounts,
        inet::ExplicitCongestionNotification,
        packet::number::{PacketNumberRange, PacketNumberSpace},
        path::Id,
        varint::{self, VarInt},
    };

    pub fn helper_new_pending(
        ack_ranges: AckRanges,
        ecn_counts: EcnCounts,
        ack_delay: Duration,
    ) -> PendingAckRanges {
        PendingAckRanges {
            ack_ranges,
            ecn_counts,
            ack_delay,
            current_active_path: None,
        }
    }

    #[test]
    fn pending_ack_ranges_test() {
        let mut now = Duration::from_millis(0);
        let mut ecn_counts = EcnCounts::default();
        let mut packet_numbers = packet_numbers_iter().step_by(2); // skip every other packet number
        let ack_ranges = AckRanges::new(3);
        let mut pending_ack_ranges = helper_new_pending(ack_ranges, ecn_counts, now);
        pending_ack_ranges.set_active_path(Id::test_id());

        assert!(pending_ack_ranges.is_empty());

        // insert range with ack_delay and ecn_counts
        now = now.saturating_add(Duration::from_millis(1));
        ecn_counts.increment(ExplicitCongestionNotification::Ect0);
        ecn_counts.increment(ExplicitCongestionNotification::Ect1);
        ecn_counts.increment(ExplicitCongestionNotification::Ce);
        let pn_a = packet_numbers.next().unwrap();
        let pn_range_a = Some(PacketNumberRange::new(pn_a, pn_a));

        assert!(pending_ack_ranges
            .extend(pn_range_a.into_iter(), Some(ecn_counts), now)
            .is_ok());

        assert_eq!(pending_ack_ranges.ack_delay, now);
        assert_eq!(pending_ack_ranges.ecn_counts, ecn_counts);
        assert!(!pending_ack_ranges.is_empty());
        assert_eq!(pending_ack_ranges.ack_ranges.interval_len(), 1);

        // insert new range with updated ack_delay and ecn_counts
        now = now.saturating_add(Duration::from_millis(1));
        ecn_counts.increment(ExplicitCongestionNotification::Ect0);
        ecn_counts.increment(ExplicitCongestionNotification::Ect1);
        ecn_counts.increment(ExplicitCongestionNotification::Ce);
        let pn_b = packet_numbers.next().unwrap();
        let pn_range_b = Some(PacketNumberRange::new(pn_b, pn_b));

        assert!(pending_ack_ranges
            .extend(pn_range_b.into_iter(), Some(ecn_counts), now)
            .is_ok());

        assert_eq!(pending_ack_ranges.ack_delay, now);
        assert_eq!(pending_ack_ranges.ecn_counts, ecn_counts);
        assert!(!pending_ack_ranges.is_empty());
        assert_eq!(pending_ack_ranges.ack_ranges.interval_len(), 2);

        // ensure pending_ack_ranges clear functionality works
        {
            assert!(!pending_ack_ranges.is_empty());
            pending_ack_ranges.reset_aggregate_info();

            assert!(pending_ack_ranges.is_empty());
            assert_eq!(pending_ack_ranges.ack_ranges.interval_len(), 0);
            assert!(!pending_ack_ranges.ack_ranges.contains(&pn_a));
            assert!(!pending_ack_ranges.ack_ranges.contains(&pn_b));
        }
    }

    #[test]
    fn iterate_range_test() {
        let now = Duration::from_millis(0);
        let ecn_counts = EcnCounts::default();
        let mut packet_numbers = packet_numbers_iter().step_by(2); // skip every other packet number
        let ack_ranges = AckRanges::new(3);
        let mut pending_ack_ranges = helper_new_pending(ack_ranges, ecn_counts, now);
        pending_ack_ranges.set_active_path(Id::test_id());

        // insert ranges
        let pn_a = packet_numbers.next().unwrap();
        let pn_range_a = Some(PacketNumberRange::new(pn_a, pn_a));
        assert!(pending_ack_ranges
            .extend(pn_range_a.into_iter(), Some(ecn_counts), now)
            .is_ok());

        let pn_b = packet_numbers.next().unwrap();
        let pn_range_b = Some(PacketNumberRange::new(pn_b, pn_b));
        assert!(pending_ack_ranges
            .extend(pn_range_b.into_iter(), Some(ecn_counts), now)
            .is_ok());

        let coll: Vec<PacketNumber> = pending_ack_ranges.iter().flatten().collect();
        assert_eq!(coll.len(), 2);
        let arr = [pn_a, pn_b];
        for pn in coll.iter() {
            assert!(arr.contains(pn));
        }
    }

    #[test]
    fn large_range_test() {
        let now = Duration::from_millis(0);
        let ecn_counts = EcnCounts::default();
        let pn_a = PacketNumberSpace::ApplicationData.new_packet_number(VarInt::from_u32(1));
        let pn_b = PacketNumberSpace::ApplicationData
            .new_packet_number(VarInt::new(varint::MAX_VARINT_VALUE).unwrap());
        let ack_ranges = AckRanges::new(3);
        let mut pending_ack_ranges = helper_new_pending(ack_ranges, ecn_counts, now);
        pending_ack_ranges.set_active_path(Id::test_id());

        let range_1 = Some(PacketNumberRange::new(pn_a, pn_b));

        assert!(pending_ack_ranges
            .extend(range_1.into_iter(), Some(ecn_counts), now)
            .is_ok());
        assert_eq!(pending_ack_ranges.ack_ranges.interval_len(), 1);
    }

    #[test]
    fn size_of_snapshots() {
        use core::mem::size_of;
        use insta::assert_debug_snapshot;

        assert_debug_snapshot!("PendingAckRanges", size_of::<PendingAckRanges>());
    }

    #[test]
    fn extend_fuzz() {
        let now = Duration::from_millis(0);
        let ecn_counts = EcnCounts::default();
        check!()
            .with_type::<(u32, u32)>()
            .map(|(a, b)| (a.min(b), a.max(b))) // ensure valid range
            .for_each(|(a, b)| {
                let ack_ranges = AckRanges::new(1);
                let mut pending_ack_ranges = helper_new_pending(ack_ranges, ecn_counts, now);
                pending_ack_ranges.set_active_path(Id::test_id());

                let pn_a = PacketNumberSpace::Initial.new_packet_number(VarInt::from_u32(*a));
                let pn_b = PacketNumberSpace::Initial.new_packet_number(VarInt::from_u32(*b));

                let range_1 = Some(PacketNumberRange::new(pn_a, pn_b));

                assert!(pending_ack_ranges
                    .extend(range_1.into_iter(), Some(ecn_counts), now)
                    .is_ok());
            });
    }
}
