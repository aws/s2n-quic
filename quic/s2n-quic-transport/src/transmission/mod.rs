// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

pub mod context;
use context::Context;

pub mod application;
pub mod connection_close;
pub mod early;

pub use crate::contexts::WriteContext;

/// re-export core
pub use s2n_quic_core::transmission::*;

use crate::{
    endpoint, path,
    space::TxPacketNumbers,
    transmission::{self, interest::Provider as _},
};
use core::{marker::PhantomData, ops::RangeInclusive};
use s2n_codec::{encoder::scatter, Encoder};
use s2n_quic_core::{
    event,
    frame::Padding,
    packet::{
        encoding::PacketPayloadEncoder,
        number::{PacketNumber, PacketNumberSpace},
        stateless_reset,
    },
    time::Timestamp,
};

pub trait Payload: interest::Provider {
    fn size_hint(&self, payload_range: RangeInclusive<usize>) -> usize;
    fn on_transmit<W: WriteContext>(&mut self, context: &mut W);
    fn packet_number_space(&self) -> PacketNumberSpace;
}

pub struct Transmission<'a, 'sub, Config: endpoint::Config, P: Payload> {
    pub config: PhantomData<Config>,
    pub outcome: &'a mut transmission::Outcome,
    pub payload: P,
    pub packet_number: PacketNumber,
    pub timestamp: Timestamp,
    pub transmission_constraint: transmission::Constraint,
    pub transmission_mode: transmission::Mode,
    pub tx_packet_numbers: &'a mut TxPacketNumbers,
    pub path_id: path::Id,
    pub publisher: &'a mut event::ConnectionPublisherSubscriber<
        'sub,
        <Config as endpoint::Config>::EventSubscriber,
    >,
    pub packet_interceptor: &'a mut <Config as endpoint::Config>::PacketInterceptor,
}

impl<'a, 'sub, Config: endpoint::Config, P: Payload> PacketPayloadEncoder
    for Transmission<'a, 'sub, Config, P>
{
    fn encoding_size_hint<E: Encoder>(&mut self, encoder: &E, minimum_len: usize) -> usize {
        if self.has_transmission_interest() {
            self.payload.size_hint(minimum_len..=encoder.capacity())
        } else {
            0
        }
    }

    fn encode(
        &mut self,
        buffer: &mut scatter::Buffer,
        minimum_len: usize,
        header_len: usize,
        tag_len: usize,
    ) {
        debug_assert!(
            buffer.is_empty(),
            "the implementation assumes an empty buffer"
        );

        let mut context: Context<Config> = Context {
            outcome: self.outcome,
            buffer,
            packet_number: self.packet_number,
            transmission_constraint: self.transmission_constraint,
            transmission_mode: self.transmission_mode,
            timestamp: self.timestamp,
            header_len,
            tag_len,
            config: Default::default(),
            path_id: self.path_id,
            publisher: self.publisher,
        };

        self.payload.on_transmit(&mut context);

        if !context.buffer.is_empty() {
            // Add padding up to minimum_len
            let mut length = minimum_len.saturating_sub(context.buffer.len());

            // if we've only got a few bytes left in the buffer may as well pad it to full
            // capacity
            let remaining_capacity = context.buffer.remaining_capacity();
            if remaining_capacity < stateless_reset::min_indistinguishable_packet_len(tag_len) {
                length = remaining_capacity;
            }

            if length > 0 {
                context.write_frame(&Padding { length });
            }

            {
                use s2n_quic_core::{
                    event::ConnectionPublisher,
                    packet::interceptor::{Interceptor, Packet},
                };

                // intercept the payload before it is encrypted
                self.packet_interceptor.intercept_tx_payload(
                    &self.publisher.subject(),
                    &Packet {
                        number: self.packet_number,
                        timestamp: self.timestamp,
                    },
                    buffer,
                );
            }

            self.tx_packet_numbers.on_transmit(self.packet_number);
            self.outcome.bytes_sent = header_len + tag_len + buffer.len();
        }
    }
}

impl<'a, 'sub, Config: endpoint::Config, P: Payload> transmission::interest::Provider
    for Transmission<'a, 'sub, Config, P>
{
    fn transmission_interest<Q: transmission::interest::Query>(
        &self,
        query: &mut Q,
    ) -> transmission::interest::Result {
        self.payload.transmission_interest(query)
    }
}
