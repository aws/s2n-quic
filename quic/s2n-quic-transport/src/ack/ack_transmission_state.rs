// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::transmission;
use s2n_quic_core::ack;

#[derive(Clone, Copy, Default, Debug, PartialEq, Eq)]
pub enum AckTransmissionState {
    /// No ACK frames will be transmitted
    #[default]
    Disabled,

    /// An ACK frame may be transmitted, but isn't required. While in this
    /// state, the ACK Manager won't express interest in transmission, but
    /// will transmit ACK frames if other components are transmitting.
    Passive {
        /// The number of remaining tranmissions for the current ack ranges
        retransmissions: usize,
    },

    /// An ACK frame is required to be transmitted. While in this state,
    /// the ACK Manager will express interest in transmission.
    Active {
        /// The number of remaining tranmissions for the current ack ranges
        retransmissions: usize,
    },
}

impl AckTransmissionState {
    /// Returns `true` if the state is set to `Active`
    pub fn is_active(&self) -> bool {
        matches!(self, Self::Active { .. })
    }

    /// Returns `true` if ACK frames should be transmitted, either actively or passively
    pub fn should_transmit(
        &self,
        constraint: transmission::Constraint,
        mode: transmission::Mode,
        has_ranges: bool,
    ) -> bool {
        match self {
            // If we don't have any ranges to transmit, then we can't do anything
            _ if !has_ranges => false,
            // Transmitting ACK frames in probes can help the peer recover faster. Since we're sending
            // a packet here, we might as well bundle it.
            _ if !mode.is_normal() => true,
            Self::Disabled => false,
            // Only transmit acks in Passive mode if we can transmit other frames as well
            Self::Passive { .. } => constraint.can_transmit() || constraint.can_retransmit(),
            //= https://www.rfc-editor.org/rfc/rfc9002#section-7
            //# packets containing only ACK frames do not count
            //# towards bytes in flight and are not congestion controlled.
            Self::Active { .. } => true,
        }
    }

    /// Transitions the transmission to active if there are pending retransmissions
    pub fn activate(&mut self) -> &mut Self {
        if let Self::Passive { retransmissions } = *self {
            *self = AckTransmissionState::Active { retransmissions }
        }

        self
    }

    /// Notify the transmission state that pending ack ranges has updated
    pub fn on_update(&mut self, ack_ranges: &ack::Ranges) -> &mut Self {
        // no need to transmit anything now
        if ack_ranges.is_empty() {
            *self = AckTransmissionState::Disabled;
            return self;
        }

        //= https://www.rfc-editor.org/rfc/rfc9000#section-13.2
        //# In general, frequent feedback from a receiver improves loss and
        //# congestion response, but this has to be balanced against excessive
        //# load generated by a receiver that sends an ACK frame in response to
        //# every ack-eliciting packet.

        // By default, ACK ranges are not retransmitted
        let mut new_retransmissions = 0;

        // A high `interval_len` indicates more RX packet reordering and/or loss
        //
        // Increment retransmissions for every N intervals we are ACKing
        //
        // TODO: 2 was chosen as an initial hunch - we need to do some experiments to improve
        const INTERVAL_SCALE: usize = 2;
        new_retransmissions += ack_ranges.interval_len() / INTERVAL_SCALE;

        // A high count indicates more outstanding RX packets to be ACKed, which
        // results in higher risk of the peer retransmitting packet data which has
        // already been processed.
        //
        // Increment retransmissions for every N packets we are ACKing to increase
        // the likelihood of ACK frames being received by the peer.
        //
        // TODO: 10 was chosen as an initial hunch - we need to do some experiments to improve
        const RANGE_SCALE: usize = 10;
        new_retransmissions += ack_ranges.spread() / RANGE_SCALE;

        /// We shouldn't retransmit too many redundant ACK frames so we'll cap it
        const MAX_RETRANSMISSIONS: usize = 10;

        new_retransmissions = new_retransmissions.min(MAX_RETRANSMISSIONS);

        match self {
            Self::Active { retransmissions } => {
                *retransmissions = new_retransmissions;
            }
            Self::Passive { retransmissions } => {
                *retransmissions = new_retransmissions;
            }
            Self::Disabled => {
                *self = AckTransmissionState::Passive {
                    retransmissions: new_retransmissions,
                };
            }
        }

        self
    }

    /// Notifies the transmission state that a transmission occurred
    pub fn on_transmit(&mut self, has_ranges: bool) -> &mut Self {
        debug_assert!(
            has_ranges,
            "ACKs can only be sent when there are ranges to send"
        );

        match *self {
            Self::Active { retransmissions } | Self::Passive { retransmissions } => {
                if let Some(retransmissions) = retransmissions.checked_sub(1) {
                    *self = AckTransmissionState::Passive { retransmissions };
                } else {
                    *self = AckTransmissionState::Disabled;
                }
            }
            Self::Disabled => {
                debug_assert!(
                    has_ranges,
                    "ACKs should not be transmitted in `Disabled` state"
                );
            }
        }

        self
    }
}

impl transmission::interest::Provider for AckTransmissionState {
    #[inline]
    fn transmission_interest<Q: transmission::interest::Query>(
        &self,
        query: &mut Q,
    ) -> transmission::interest::Result {
        if self.is_active() {
            //= https://www.rfc-editor.org/rfc/rfc9002#section-7
            //# packets containing only ACK frames do not count
            //# towards bytes in flight and are not congestion controlled
            query.on_forced()?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::{super::tests::packet_numbers_iter, *};
    use insta::assert_debug_snapshot;

    #[test]
    fn should_transmit_test() {
        let mut results = vec![];
        for state in &[
            AckTransmissionState::Disabled,
            AckTransmissionState::Passive { retransmissions: 1 },
            AckTransmissionState::Active { retransmissions: 1 },
        ] {
            for constraint in &[
                transmission::Constraint::None,
                transmission::Constraint::AmplificationLimited,
                transmission::Constraint::CongestionLimited,
                transmission::Constraint::RetransmissionOnly,
            ] {
                for mode in &[
                    transmission::Mode::Normal,
                    transmission::Mode::MtuProbing,
                    transmission::Mode::PathValidationOnly,
                    transmission::Mode::LossRecoveryProbing,
                ] {
                    for has_ranges in &[true, false] {
                        let res = (
                            state,
                            *constraint,
                            *mode,
                            *has_ranges,
                            state.should_transmit(*constraint, *mode, *has_ranges),
                        );
                        results.push(res);
                    }
                }
            }
        }

        assert_debug_snapshot!("should_transmit", results);
    }

    #[test]
    fn activate_test() {
        assert!(
            !AckTransmissionState::Disabled.activate().is_active(),
            "disabled state should not activate"
        );
        assert!(
            AckTransmissionState::Passive { retransmissions: 1 }
                .activate()
                .is_active(),
            "passive state should activate"
        );
        assert!(
            AckTransmissionState::Active { retransmissions: 1 }
                .activate()
                .is_active(),
            "active state should activate"
        );
    }

    #[test]
    #[should_panic]
    fn disabled_transmission_test() {
        AckTransmissionState::Disabled.on_transmit(false);
    }

    #[test]
    fn transmission_test() {
        assert_eq!(
            *AckTransmissionState::Passive { retransmissions: 0 }.on_transmit(true),
            AckTransmissionState::Disabled,
            "transmitting should transition to Disabled"
        );

        assert_eq!(
            *AckTransmissionState::Passive { retransmissions: 1 }.on_transmit(true),
            AckTransmissionState::Passive { retransmissions: 0 },
            "transmitting should decrement and stay in the same state"
        );

        assert_eq!(
            *AckTransmissionState::Disabled.on_transmit(true),
            AckTransmissionState::Disabled,
            "transmitting when disabled should continue to be disabled"
        );
    }

    #[test]
    fn update_test() {
        let mut ack_ranges = ack::Ranges::new(10);
        let mut packet_numbers = packet_numbers_iter().step_by(2); // skip every other packet number

        assert_eq!(
            *AckTransmissionState::Passive { retransmissions: 1 }.on_update(&ack_ranges),
            AckTransmissionState::Disabled,
            "empty ack_ranges should transition to Disabled"
        );

        assert!(ack_ranges
            .insert_packet_number(packet_numbers.next().unwrap())
            .is_ok());

        assert_eq!(
            *AckTransmissionState::Disabled.on_update(&ack_ranges),
            AckTransmissionState::Passive { retransmissions: 0 },
            "one ack range should transition to passive"
        );

        assert!(ack_ranges
            .insert_packet_number(packet_numbers.next().unwrap())
            .is_ok());
        assert!(ack_ranges
            .insert_packet_number(packet_numbers.next().unwrap())
            .is_ok());

        assert_eq!(
            *AckTransmissionState::Disabled.on_update(&ack_ranges),
            AckTransmissionState::Passive { retransmissions: 1 },
            "multiple ack ranges should transition to passive with retransmissions"
        );

        assert_eq!(
            *AckTransmissionState::Passive { retransmissions: 0 }.on_update(&ack_ranges),
            AckTransmissionState::Passive { retransmissions: 1 },
            "multiple ack ranges should update passive with retransmissions"
        );

        assert_eq!(
            *AckTransmissionState::Active { retransmissions: 0 }.on_update(&ack_ranges),
            AckTransmissionState::Active { retransmissions: 1 },
            "multiple ack ranges should update active with retransmissions"
        );
    }

    #[test]
    #[cfg(target_pointer_width = "64")]
    fn size_of_snapshots() {
        use core::mem::size_of;
        use insta::assert_debug_snapshot;

        assert_debug_snapshot!("AckTransmissionState", size_of::<AckTransmissionState>());
    }
}
