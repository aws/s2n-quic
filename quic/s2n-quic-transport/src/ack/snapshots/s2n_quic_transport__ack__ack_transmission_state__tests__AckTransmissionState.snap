---
source: quic/s2n-quic-transport/src/ack/ack_transmission_state.rs
expression: "size_of::<AckTransmissionState>()"
---
16
